### PR TITLE
Support form submission of multipart/form-data

### LINKED ISSUE #7553: Add support for multipart/form-data
This lets files be submitted via XHR or forms.

There's a parser [here](https://github.com/mikedilger/formdata), which we can use. I believe we already have enough Blob/File support for this, we just need to wire things up.

We should just mirror URLSearchParams and implement a [`serialize()`](https://github.com/servo/servo/blob/71b277d5675556e61a82ae9dbf3105449c3a8275/components/script/dom/urlsearchparams.rs#L129) on `FormData`, the rest of the plumbing for XHR should mostly be done already.

For forms, we'll need form upload control support, which we can try later.

### PR DESCRIPTION
**Pull request by KiChjang**
Fixes #7553.

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/10712)

### CODE REVIEW

**Comment by jdm** [comment]
Looks like no WPT tests changed results; what's the test plan here?

**Comment by KiChjang** [comment]
I'm _very_ surprised that no new WPT tests pass, I'm guessing I must be doing something wrong in this PR.

**Comment by KiChjang** [comment]
Ah wait, it's because there aren't any consumers of the functionality yet.

**Comment by KiChjang** [comment]
I'm tempted to also implement submit as body action for forms just to make tests for this, but this'll have to wait until saturday or later since I have final exams and I have some moving to do.

File: tests/wpt/web-platform-tests/html/semantics/forms/form-submission-0/resources/url-encoded.py [hunk 1]
```diff
@@ -0,0 +1,5 @@
+def main(request, response):
+    if request.body == "foo=bara":
+        return [("Content-Type", "text/plain")], "OK"
+    else:
+        return [("Content-Type", "text/plain")], "FAIL"
```
Comment (KiChjang) [inline tests/wpt/web-platform-tests/html/semantics/forms/form-submission-0/resources/url-encoded.py]: This doesn't work right now - it keeps on returning "Unknown content type (application/json)" to the iframe. Not sure what's wrong here...

**Comment by Ms2ger** [comment]
Review status: 0 of 11 files reviewed at latest revision, 3 unresolved discussions.

---

*[tests/wpt/web-platform-tests/html/semantics/forms/form-submission-0/submit-entity-body.html, line 21 \[r5\]](https://reviewable.io:443/reviews/servo/servo/10712#-KGC1Hj2_EZHxO4RERky:-KGC1Hj3qh9MwfzIfM6o:-473090924) ([raw file](https://github.com/servo/servo/blob/04fac3a27bdcf160cb9295d1316f8d592e3a2eaa/tests/wpt/web-platform-tests/html/semantics/forms/form-submission-0/submit-entity-body.html#L21)):*
`var`

---

*[tests/wpt/web-platform-tests/html/semantics/forms/form-submission-0/submit-entity-body.html, line 39 \[r5\]](https://reviewable.io:443/reviews/servo/servo/10712#-KGC1KtUbKccMzs8UgLL:-KGC1KtUbKccMzs8UgLM:-1902112513) ([raw file](https://github.com/servo/servo/blob/04fac3a27bdcf160cb9295d1316f8d592e3a2eaa/tests/wpt/web-platform-tests/html/semantics/forms/form-submission-0/submit-entity-body.html#L39)):*
I don't think it's guaranteed that this line of JS will execute before the `load` event has been dispatched; try putting the `iframe` after the script and using the `onload` content attribute in markup.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10712)*

File: components/compositing/constellation.rs [hunk 2]
```diff
@@ -1008,13 +1008,19 @@ impl<LTF: LayoutThreadFactory, STF: ScriptThreadFactory> Constellation<LTF, STF>
...
 
+        let load_data = if let Some(mut data) = load_info.load_data {
+            data.url = new_url;
+            data
+        } else {
+            LoadData::new(new_url, None, None)
+        };
+
         // Create the new pipeline, attached to the parent and push to pending frames
-        // TODO - loaddata here should have referrer info (not None, None)
```
Comment (jdm) [inline components/compositing/constellation.rs]: This comment should be maintained a couple lines above.

File: components/script/dom/htmliframeelement.rs [hunk 3]
```diff
@@ -148,7 +148,7 @@ impl HTMLIFrameElement {
             None => Url::parse("about:blank").unwrap(),
         };
 
-        self.navigate_or_reload_child_browsing_context(Some(url));
+        self.navigate_or_reload_child_browsing_context(Some(LoadData::new(url, None, None)));
```
Comment (jdm) [inline components/script/dom/htmliframeelement.rs]: Add a comment about adding a proper referrer policy and referrer URL.

**Comment by KiChjang** [comment]
Comments addressed. r? @jdm

**Comment by jdm** [comment]
This is good to merge when rebased.

**Comment by KiChjang** [comment]
@bors-servo retry

* Test for intermittent

**Pull request merged**